src/vs/workbench/services/search/common/textSearchManager.ts

```diff
@@ -0,0 +1,355 @@
+		const tokenSource = new CancellationTokenSource();
+		token.onCancellationRequested(() => tokenSource.cancel());
+
+		return new Promise<ISearchCompleteStats>((resolve, reject) => {
+			this.collector = new TextSearchResultsCollector(onProgress);
+
+			let isCanceled = false;
+			const onResult = (result: TextSearchResult, folderIdx: number) => {
+				if (isCanceled) {
+					return;
+				}
+
+				if (!this.isLimitHit) {
+					const resultSize = this.resultSize(result);
+					if (extensionResultIsMatch(result) && typeof this.query.maxResults === 'number' && this.resultCount + resultSize > this.query.maxResults) {
+						this.isLimitHit = true;
+						isCanceled = true;
+						tokenSource.cancel();
+
+						result = this.trimResultToSize(result, this.query.maxResults - this.resultCount);
+					}
+
+					const newResultSize = this.resultSize(result);
+					this.resultCount += newResultSize;
+					if (newResultSize > 0) {
+						this.collector!.add(result, folderIdx);
+					}
+				}
+			};
+
+			// For each root folder
+			Promise.all(folderQueries.map((fq, i) => {
+				return this.searchInFolder(fq, r => onResult(r, i), tokenSource.token);
+			})).then(results => {
+				tokenSource.dispose();
+				this.collector!.flush();
+
+				const someFolderHitLImit = results.some(result => !!result && !!result.limitHit);
+				resolve({
+					limitHit: this.isLimitHit || someFolderHitLImit,
+					stats: {
+						type: 'textSearchProvider'
+					}
+				});
+			}, (err: Error) => {
+				tokenSource.dispose();
+				const errMsg = toErrorMessage(err);
+				reject(new Error(errMsg));
+			});
+		});
+	}
+
+	private resultSize(result: TextSearchResult): number {
+		const match = <TextSearchMatch>result;
+		return Array.isArray(match.ranges) ?
+			match.ranges.length :
+			1;
+	}
+
+	private trimResultToSize(result: TextSearchMatch, size: number): TextSearchMatch {
+		const rangesArr = Array.isArray(result.ranges) ? result.ranges : [result.ranges];
+		const matchesArr = Array.isArray(result.preview.matches) ? result.preview.matches : [result.preview.matches];
+
+		return {
+			ranges: rangesArr.slice(0, size),
+			preview: {
+				matches: matchesArr.slice(0, size),
+				text: result.preview.text
+			},
+			uri: result.uri
+		};
+	}
+
+	private searchInFolder(folderQuery: IFolderQuery<URI>, onResult: (result: TextSearchResult) => void, token: CancellationToken): Promise<TextSearchComplete | null | undefined> {
+		const queryTester = new QueryGlobTester(this.query, folderQuery);
+		const testingPs: Promise<void>[] = [];
+		const progress = {
+			report: (result: TextSearchResult) => {
+				if (!this.validateProviderResult(result)) {
+					return;
+				}
+
+				const hasSibling = folderQuery.folder.scheme === 'file' ?
+					glob.hasSiblingPromiseFn(() => {
+						return this.fileUtils.readdir(resources.dirname(result.uri));
+					}) :
+					undefined;
+
+				const relativePath = resources.relativePath(folderQuery.folder, result.uri);
```

Comment:
@roblourens note that this used to be `path.relative(result.uri.fsPath)`. I changed it to use `URI` instead.